### PR TITLE
add fill_diagonal function

### LINKED ISSUE #21796: Add torch.fill_diagonal
## 🚀 Feature
Implement torch.fill_diagonal based on numpy.fill_diagonal

See https://docs.scipy.org/doc/numpy/reference/generated/numpy.fill_diagonal.html

## Motivation
Alex asked me how to do this today.

## Alternatives
The following works but isn't obvious and may not work for non-strided Tensors:

```python
torch.diagonal(x).fill_(value)
```

### PR DESCRIPTION
**Pull request by shihongzhi**
Fixes #21796

### CODE REVIEW

**Review by colesbury — CHANGES_REQUESTED** [review]
Thanks for taking on this issue! This needs a few changes before it can be merged:

1) The API should match the NumPy API: https://docs.scipy.org/doc/numpy/reference/generated/numpy.fill_diagonal.html
2) The function needs API documentation (in [torch/_torch_docs.py](https://github.com/pytorch/pytorch/blob/master/torch/_torch_docs.py))
3) The function needs unit tests. (In test/test_torch.py)

Once the PR is ready for review, please click "dismiss review" on this

File: aten/src/ATen/native/TensorFactories.cpp
Comment (SsnL) [inline aten/src/ATen/native/TensorFactories.cpp]: doesn't tensor.diagonal already do these checks?

Comment (shihongzhi) [reply aten/src/ATen/native/TensorFactories.cpp]: yes. I remove these checks, and change api to match Numpy api. Thank you

**Comment by apaszke** [comment]
To add to what @colesbury said, the name will also need to be changed to have a trailing underscore, since it modifies the input tensor.

**Comment by shihongzhi** [comment]
@apaszke  Already add trailing underscore to function name. Thank you

@colesbury  I had add docs&unit tests and match this API to Numpy API

File: aten/src/ATen/native/native_functions.yaml
Comment (colesbury) [inline aten/src/ATen/native/native_functions.yaml]: Do we want this to be a function on torch? Originally, we only had in-place operations as methods on torch.Tensor but we've seemed to relax that (perhaps unintentionally)

cc @gchanan @apaszke

File: test/test_torch.py
Comment (colesbury) [inline test/test_torch.py]: Please add test cases for:

1) A 3-d or larger tensor
2) A non-contiguous tensor

File: torch/_torch_docs.py
Comment (colesbury) [inline torch/_torch_docs.py]: wrap (bool):

(add type)

File: torch/_torch_docs.py
Comment (colesbury) [inline torch/_torch_docs.py]: return -> returns

File: aten/src/ATen/native/TensorFactories.cpp
Comment (colesbury) [inline aten/src/ATen/native/TensorFactories.cpp]: large -> larger

remove trailing space after `1`

**Review by colesbury — COMMENTED** [review]
Nice! Looking pretty good

File: aten/src/ATen/native/native_functions.yaml
Comment (gchanan) [reply aten/src/ATen/native/native_functions.yaml]: I like in-place ops being method only.

Comment (shihongzhi) [reply aten/src/ATen/native/native_functions.yaml]: I remove 'function'

Comment (apaszke) [reply aten/src/ATen/native/native_functions.yaml]: Yes, it should only be a method.

File: torch/_tensor_docs.py
Comment (gchanan) [inline torch/_tensor_docs.py]: I'm having a bit of trouble parsing the description -- do you mean "Fill the main diagonal of a tensor that has at least 2-dimensions?"

Comment (shihongzhi) [reply torch/_tensor_docs.py]: sorry, the description is so confused. I'm changed this.

**Comment by shihongzhi** [comment]
@colesbury I fixed those bug. It's ready for review again.

File: docs/source/torch.rst
Comment (colesbury) [inline docs/source/torch.rst]: I think this needs to be removed since it's no longer on the torch package

File: test/test_torch.py
Comment (colesbury) [inline test/test_torch.py]: please remove the extra whitespace lines

**Review by colesbury — CHANGES_REQUESTED** [review]
Some of the continuous builds are failing (`test_torch (__main__.TestDocCoverage)`) I think because of the function listed in torch.rst but it's not in the torch package.

Otherwise looks good to go

**Comment by shihongzhi** [comment]
@colesbury Thank you.  I removed `fill_diagonal_` in torch.rst

**Comment by shihongzhi** [comment]
@colesbury I think this's ready for merge

**Comment by colesbury** [comment]
Thanks @shihongzhi!

**Review by colesbury — APPROVED** [review]

**Comment by xuhdev** [comment]
Can we move the definition to a different file? It seems to me that fill_diagonal is not a factory. Perhaps create Fill.cpp and move both `fill_diagonal` and `fill` to it? I don't think this is urgent for this PR -- we can do this after this PR is merged.

**Comment by colesbury** [comment]
@xuhdev yes, that seems reasonable. Do you want to make the change after this lands?

**Comment by xuhdev** [comment]
@colesbury See #22758

**Pull request closed without merging**
